File: algo_2021/w5.-traverse-divide-and-conquer/93.Balanced_Binary_tree/93_balanced-binary_tree.py
"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""

class Solution:
    """
    @param root: The root of binary tree.
    @return: True if this Binary tree is Balanced, or false.
    """
    def isBalanced(self, root):
        # write your code here
        is_balanced, _ = self.validate(root)
        return is_balanced
    
    def validate(self, root):
        # if empty tree, it's balanced and return heigh = 0
        if not root:
            return True, 0
        
        is_left_balanced, left_height = self.validate(root.left)
        is_right_balanced, right_height = self.validate(root.right)
        root_height = max(left_height, right_height) + 1

        if not is_left_balanced or not is_right_balanced: 
            return False, root_height
        if abs(left_height - right_height) > 1:
            return False, root_height
        return True, root_height
        

File: algo_2021/w5.-traverse-divide-and-conquer/93.Balanced_Binary_tree/93_balanced-binary_tree.java
/**
 * Definition of TreeNode:
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left, right;
 *     public TreeNode(int val) {
 *         this.val = val;
 *         this.left = this.right = null;
 *     }
 * }
 */


public class Solution {
    /**
     * @param root: The root of binary tree.
     * @return: True if this Binary tree is Balanced, or false.
     */
    class Res {
    public boolean isBalanced;
    public int height;
    public Res(boolean isBalanced, int height) {
        this.isBalanced = isBalanced;
        this.height = height;
    }
}
    
    public boolean isBalanced(TreeNode root) {
        // write your code here
        Res result= validate(root);
        return result.isBalanced;
    }

    private Res validate(TreeNode root) {
        if (root == null) {
            return new Res(true, 0);
        }

        Res leftRes = validate(root.left);
        Res rightRes = validate(root.right);
        int rootHeight = Math.max(leftRes.height, rightRes.height) + 1;

        if (!leftRes.isBalanced || !rightRes.isBalanced) {
            return new Res(false, rootHeight);
        }
        if (Math.abs(leftRes.height - rightRes.height) > 1) {
            return new Res(false, rootHeight);
        }
        return new Res(true, rootHeight);
        
    }
}